src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1040,49 +1038,73 @@ private enum AssignmentKind
         {
+            // Callers should apply any conversions before calling this method
+            // (see https://github.com/dotnet/roslyn/issues/39867).
+            if (targetType.HasType &&
+                !targetType.Type.Equals(valueType.Type, TypeCompareKind.AllIgnoreOptions))
+            {
+                return false;
+            }
+
             if (value == null ||
                 value.WasCompilerGenerated ||
                 !targetType.HasType ||
-                targetType.Type.IsValueType ||
-                targetType.CanBeAssignedNull ||
-                valueType.IsNotNull)
+                targetType.Type.IsValueType)
             {
                 return false;
             }
 
-            location ??= value.Syntax.GetLocation();
-            var unwrappedValue = SkipReferenceConversions(value);
-            if (unwrappedValue.IsSuppressed)
+            switch (targetType.NullableAnnotation)
             {
-                return false;
+                case NullableAnnotation.Oblivious:
+                case NullableAnnotation.Annotated:
+                    return false;
             }
 
-            if (RequiresSafetyWarningWhenNullIntroduced(targetType))
+            switch (valueType.State)
             {
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                if (conversion.Kind == ConversionKind.UnsetConversionKind)
-                    conversion = this._conversions.ClassifyImplicitConversionFromType(valueType.Type, targetType.Type, ref useSiteDiagnostics);
-
-                if (conversion.IsImplicit && !conversion.IsDynamic)
-                {
-                    // For type parameters that cannot be annotated, the analysis must report those
-                    // places where null values first sneak in, like `default`, `null`, and `GetFirstOrDefault`,
-                    // as a safety diagnostic.  This is NOT one of those places.
+                case NullableFlowState.NotNull:
                     return false;
-                }
+                case NullableFlowState.MaybeNull:
+                    if (targetType.Type.IsTypeParameterDisallowingAnnotation())
+                    {
+                        return false;
+                    }
+                    break;
+            }
 
-                useLegacyWarnings = false;
+            location ??= value.Syntax.GetLocation();
+            var unwrappedValue = SkipReferenceConversions(value);
+            if (unwrappedValue.IsSuppressed)
+            {
+                return false;
             }
 
-            if (reportNullLiteralAssignmentIfNecessary(value, location, valueType.ToTypeWithAnnotations()))
+            if (useLegacyWarnings &&
+                valueType.Type?.TypeKind == TypeKind.TypeParameter &&
+                valueType.State == NullableFlowState.MaybeDefault)
             {
+                // No W warning reported assigning or casting [MaybeNull]T value to T
```

Comment:
>No [](start = 19, length = 2)

OK, but can you please add some documentation about what the return value of this method means?  Something like "Returns true if there was a nullable mismatch between the assigned value and the assigned location."

Reply:
Good point. I've changed the method to `void` to avoid confusion.

---
In reply to: [347681909](https://github.com/dotnet/roslyn/pull/39778#discussion_r347681909) [](ancestors = 347681909)